src/components/authoring/authoring-app.tsx

```diff
@@ -0,0 +1,183 @@
+                        <Button label="Remove" onClick={this.removeDef.bind(this, def.word)}/>
+                      </td>
+                    </tr>;
+                  }
+                })
+              }
+            </tbody>
+          </table>
+          {!newDefEditor && <Button icon="iconPlus" label="Add a new definition" onClick={this.toggleNewDef}/>}
+          {
+            newDefEditor &&
+            <DefinitionEditor
+              onCancel={this.toggleNewDef}
+              onSave={this.addNewDef}
+            />
+          }
+        </div>
+        <div className={css.jsonSection}>
+          <h2>Glossary JSON</h2>
+          <p><Button label="Copy JSON to Clipboard" onClick={this.copyJSON} /></p>
+          <div className={css.help}>
+            Note that the editor below accepts and displays JS object syntax instead of the JSON notation.
+            Always use button above to copy correctly formatted JSON string.
+          </div>
+          <JSONEditor
+            placeholder={glossary}
+            onChange={this.handleJSONChange}
+            width="600px"
+            height="400px"
+          />
+        </div>
+        <div className={css.preview}>
+          <h2>Preview</h2>
+          <div className={css.handle}>
+            <span className={icons.iconBook}/>
+            <div>Glossary</div>
+          </div>
+          <div className={css.sidebar}>
+            <GlossarySidebar
+              definitions={definitions}
+              learnerDefinitions={{}}
+            />
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  private copyJSON = () => {
+    const { glossary } = this.state;
+    const fakeInput = document.createElement("textarea");
+    fakeInput.value = JSON.stringify(glossary, null, 2);
+    document.body.appendChild(fakeInput);
+    fakeInput.select();
+    document.execCommand("copy");
+    document.body.removeChild(fakeInput);
+  }
+
+  private handleJSONChange = (data: IGlossary) => {
+    this.setState({ glossary: data });
+  }
+
+  private handleAskForUserDefChange = (event: React.ChangeEvent) => {
+    const glossary: IGlossary = clone(this.state.glossary);
+    glossary.askForUserDefinition = (event.target as HTMLInputElement).checked;
+    this.setState({ glossary });
+  }
+
+  private toggleDefinitionEditor = (word: string) => {
+    const definitionEditors = clone(this.state.definitionEditors);
+    definitionEditors[word] = !definitionEditors[word];
+    this.setState({ definitionEditors });
+  }
+
+  private toggleNewDef = () => {
+    const { newDefEditor } = this.state;
+    this.setState({ newDefEditor: !newDefEditor });
+  }
+
+  private editDef = (newDef: IWordDefinition) => {
+    const glossary: IGlossary = clone(this.state.glossary);
+    const definitionEditors = clone(this.state.definitionEditors);
+    const existingDefIdx = glossary.definitions.map(d => d.word).indexOf(newDef.word);
+    if (existingDefIdx !== -1) {
```

Comment:
I wonder what we do if we don't find the word in the index? It's an error, but will it ever  happen? 
 
The presence of the `if` check seems to indicate that it's a possibility we need to consider. If so, shouldn't we be explicit about what do if we don't find it?

Reply:
You're right, that should never happen. I'll remove this `if`. If it happens, it means that something went really bad and things should explode. :)